source/extensions/http/custom_response/redirect_policy/redirect_policy.cc

```diff
@@ -0,0 +1,175 @@
+      modify_request_headers_action_(createModifyRequestHeadersAction(config, context)) {
+  ::Envoy::Http::Utility::Url absolute_url;
+  const std::string uri(absl::StrCat(host_, path_));
+  if (!absolute_url.initialize(uri, false)) {
+    throw EnvoyException(
+        absl::StrCat("Invalid uri specified for redirection for custom response: ", uri));
+  }
+}
+
+std::unique_ptr<ModifyRequestHeadersAction> RedirectPolicy::createModifyRequestHeadersAction(
+    const envoy::extensions::http::custom_response::redirect_policy::v3::RedirectPolicy& config,
+    Envoy::Server::Configuration::ServerFactoryContext& context) {
+  std::unique_ptr<ModifyRequestHeadersAction> action;
+  if (config.has_modify_request_headers_action()) {
+    auto& factory = Envoy::Config::Utility::getAndCheckFactory<ModifyRequestHeadersActionFactory>(
+        config.modify_request_headers_action());
+    const auto action_config = Envoy::Config::Utility::translateAnyToFactoryConfig(
+        config.modify_request_headers_action().typed_config(), context.messageValidationVisitor(),
+        factory);
+    action = factory.createAction(*action_config, context);
+  }
+  return action;
+}
+
+::Envoy::Http::FilterHeadersStatus RedirectPolicy::encodeHeaders(
+    ::Envoy::Http::ResponseHeaderMap& headers, bool,
+    Extensions::HttpFilters::CustomResponse::CustomResponseFilter& custom_response_filter) const {
+  // If filter state for custom response exists, it means this response is a
+  // custom response. Apply the custom response mutations to the response from
+  // the remote source and return.
+  auto encoder_callbacks = custom_response_filter.encoderCallbacks();
+  auto decoder_callbacks = custom_response_filter.decoderCallbacks();
+  auto filter_state = encoder_callbacks->streamInfo().filterState()->getDataReadOnly<Policy>(
+      "envoy.filters.http.custom_response");
+  if (filter_state) {
+    ENVOY_BUG(filter_state == this, "Policy filter state should be this policy.");
+    //  Apply mutations if this is a non-error response. Else leave be.
+    auto const cr_code = ::Envoy::Http::Utility::getResponseStatusOrNullopt(headers);
+    if (!cr_code.has_value() || (*cr_code < 100 || *cr_code > 299)) {
+      return ::Envoy::Http::FilterHeadersStatus::Continue;
+    }
+    // Apply header mutations.
+    response_header_parser_->evaluateHeaders(headers, encoder_callbacks->streamInfo());
+    // Modify response status code.
+    if (status_code_.has_value()) {
+      auto const code = *status_code_;
+      headers.setStatus(std::to_string(enumToInt(code)));
+      encoder_callbacks->streamInfo().setResponseCode(static_cast<uint32_t>(code));
+    }
+    return ::Envoy::Http::FilterHeadersStatus::Continue;
+  }
+
+  // Modify the request headers & recreate stream.
+  // TODO(pradeepcrao): Currently Custom Response is not compatible with sendLocalReply.
+  if (custom_response_filter.onLocalRepplyCalled()) {
```

Comment:
I still need to process in what conditions this will trigger. We can discuss this in-person.

Reply:
Added a comment.